pkg/apis/core/v1beta1/helper/helpers_test.go

```diff
@@ -595,365 +663,427 @@ var _ = Describe("helper", func() {
+							},
+						},
+					}
+					ok, newVersion, err := GetKubernetesVersionForPatchUpdate(&cloudProfile, currentVersion)
+					Expect(err).ToNot(HaveOccurred())
+					Expect(ok).To(Equal(qualifyingVersionFound))
+					Expect(newVersion).To(Equal(expectedVersion))
+				},
+				Entry("Do not consider preview versions for patch update.",
+					"1.12.2",
+					[]gardencorev1beta1.ExpirableVersion{
+						{Version: "1.15.1"},
+						{Version: "1.15.0"},
+						{Version: "1.14.4"},
+						{
+							Version:        "1.12.9",
+							Classification: &previewClassification,
+						},
+						{
+							Version:        "1.12.4",
+							Classification: &previewClassification,
+						},
+						// latest qualifying version for updating version 1.12.2
+						{Version: "1.12.3"},
+						{Version: "1.12.2"},
+					},
+					"1.12.3",
+					true,
+				),
+				Entry("Do not consider expired versions for patch update.",
+					"1.12.2",
+					[]gardencorev1beta1.ExpirableVersion{
+						{Version: "1.15.1"},
+						{Version: "1.15.0"},
+						{Version: "1.14.4"},
+						{
+							Version:        "1.12.9",
+							ExpirationDate: &expirationDateInThePast,
+						},
+						{
+							Version:        "1.12.4",
+							ExpirationDate: &expirationDateInThePast,
+						},
+						// latest qualifying version for updating version 1.12.2
+						{Version: "1.12.3"},
+						{Version: "1.12.2"},
+					},
+					"1.12.3",
+					true,
+				),
+				Entry("Should not find qualifying version - no higher version available that is not expired or in preview.",
+					"1.12.2",
+					[]gardencorev1beta1.ExpirableVersion{
+						{Version: "1.15.1"},
+						{Version: "1.15.0"},
+						{Version: "1.14.4"},
+						{
+							Version:        "1.12.9",
+							ExpirationDate: &expirationDateInThePast,
+						},
+						{
+							Version:        "1.12.4",
+							Classification: &previewClassification,
+						},
+						{Version: "1.12.2"},
+					},
+					"",
+					false,
+				),
+				Entry("Should not find qualifying version - is already highest version of minor.",
+					"1.12.2",
+					[]gardencorev1beta1.ExpirableVersion{
+						{Version: "1.15.1"},
+						{Version: "1.15.0"},
+						{Version: "1.14.4"},
+						{Version: "1.12.2"},
+					},
+					"",
+					false,
+				),
+				Entry("Should not find qualifying version - is already on latest version of latest minor.",
+					"1.15.1",
+					[]gardencorev1beta1.ExpirableVersion{
+						{Version: "1.15.1"},
+						{Version: "1.15.0"},
+						{Version: "1.14.4"},
+						{Version: "1.12.2"},
+					},
+					"",
+					false,
+				),
+			)
 
-	Describe("#ShootExpirableVersionExists", func() {
-		var (
-			constraint        gardencorev1beta1.MachineImage
-			shootMachineImage gardencorev1beta1.ShootMachineImage
-		)
-		BeforeEach(func() {
-			constraint = gardencorev1beta1.MachineImage{
-				Name: "coreos",
-				Versions: []gardencorev1beta1.ExpirableVersion{
-					{
-						Version: "0.0.2",
+			DescribeTable("#GetKubernetesVersionForMinorUpdate",
```

Comment:
Can you add some cases here and also for machine images and predicates that lesser versions are not considered?

Reply:
Don't you think that is already tested here? We are handing over a set of versions in a CloudProfile and expect the latest qualifying version to be returned. So that is implicitly higher than a lower version of the shoot. Though I ll add a test for the FilterForSameMajorMinorVersion() and FilterForConsecutiveMinor() predicates.

Reply:
Sorry, I had to think it through again. Yes, for `GetKubernetesVersionForMinorUpdate` and `GetKubernetesVersionForPatchUpdate` you are doing the check via semver constraint checks (`~`/`^`) in the Predicates and that is now tested, perfect :).
For machine image versions you are not considering the current version in the helper, so thats also fine, no need for a check here.

But if ISIC the shoot maintenance controller will actually downgrade the shoots machine image the current image is expired and only lesser version exist, that are not expired.
For example this test for the maintenance controller, will fail:
```go
It("should not downgrade - even if highest version is expired", func() {
	highestExpiredVersion := gardencorev1beta1.ExpirableVersion{
		Version:        "1.1.2",
		Classification: &deprecatedClassification,
		ExpirationDate: &expirationDateInThePast,
	}
	cloudProfile.Spec.MachineImages[0].Versions = append(cloudProfile.Spec.MachineImages[0].Versions, highestExpiredVersion)
	shoot.Spec.Provider.Workers[0].Machine.Image.Version = &highestExpiredVersion.Version
	workerImages, _, err := MaintainMachineImages(testlogger, shoot, cloudProfile)

	Expect(err).To(BeNil())
	Expect(len(workerImages)).NotTo(Equal(0))
	Expect(workerImages[0].Name).To(Equal(cloudProfile.Spec.MachineImages[0].Name))
	Expect(*workerImages[0].Version).To(Equal("1.1.2"))
})
```

Reply:
Thanks for taking the time and checking. Yes when the CloudProfile is misconfigured that can happen and should be prevented.  I added a check.

I can see from your code that you would not expect an error in that scenario. While I think that there should be no maintenance error event on the Shoot (nothing the Shoot can do about it), I think the reconcile should throw and log an error to be retried eventually. The operator would need to step in and provide a proper version. 

WDYT

Reply:
Actually, the code was just for demonstrating what I meant, haven't given it so much thought ;) 
Yeah, I think at least an error log would be appropriate for that case, but no strong opinion here.